invoice/invoice.go

```diff
@@ -0,0 +1,938 @@
+	return func(i *Invoice) {
+		i.Description = &description
+	}
+}
+
+// DescriptionHash is a functional option that allows callers of NewInvoice to
+// set the payment description hash of the created Invoice.
+// Note: Must be used if and only if Description is not used.
+func DescriptionHash(descriptionHash [32]byte) func(*Invoice) {
+	return func(i *Invoice) {
+		i.DescriptionHash = &descriptionHash
+	}
+}
+
+// Expiry is a functional option that allows callers of NewInvoice to set the
+// expiry of the created Invoice. If not set, a default expiry of 60 min will
+// be implied.
+func Expiry(expiry time.Time) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Expiry = &expiry
+	}
+}
+
+// FallbackAddr is a functional option that allows callers of NewInvoice to set
+// the Invoice's fallback on-chain address that can be used for payment in case
+// the Lightning payment fails
+func FallbackAddr(fallbackAddr btcutil.Address) func(*Invoice) {
+	return func(i *Invoice) {
+		i.FallbackAddr = fallbackAddr
+	}
+}
+
+// RoutingInfo is a functional option that allows callers of NewInvoice to set
+// one or more entries containing extra routing information for a private route
+// to the target node.
+func RoutingInfo(routingInfo []ExtraRoutingInfo) func(*Invoice) {
+	return func(i *Invoice) {
+		i.RoutingInfo = routingInfo
+	}
+}
+
+// NewInvoice creates a new Invoice object. The last parameter is a set of
+// variadic argumements for setting optional fields of the invoice.
+// Note: Either Description  or DescriptionHash must be provided for the Invoice
+// to be considered valid.
+func NewInvoice(net *chaincfg.Params, paymentHash [32]byte,
+	timestamp time.Time, options ...func(*Invoice)) (*Invoice, error) {
+
+	invoice := &Invoice{
+		Net:         net,
+		PaymentHash: &paymentHash,
+		Timestamp:   timestamp,
+	}
+
+	for _, option := range options {
+		option(invoice)
+	}
+
+	if err := validateInvoice(invoice); err != nil {
+		return nil, err
+	}
+
+	return invoice, nil
+}
+
+// Decode parses the provided encoded invoice, and returns a decoded Invoice in
+// case it is valid by BOLT-0011.
+func Decode(invoice string) (*Invoice, error) {
+	decodedInvoice := Invoice{}
+
+	// Decode the invoice using the modified bech32 decoder.
+	hrp, data, err := decodeBech32(invoice)
+	if err != nil {
+		return nil, err
+	}
+
+	// We expect the human-readable part to at least have ln + two chars
+	// encoding the network.
+	if len(hrp) < 4 {
+		return nil, fmt.Errorf("hrp too short")
+	}
+
+	// First two characters of HRP should be "ln".
+	if hrp[:2] != "ln" {
+		return nil, fmt.Errorf("prefix should be \"ln\"")
+	}
+
+	// The next two characters should be a valid prefix for a segwit BIP173
+	// address.
+	segwitPrefix := hrp[2:4] + "1"
+	if !chaincfg.IsBech32SegwitPrefix(segwitPrefix) {
+		return nil, fmt.Errorf("unknown network \"%v\"", hrp[2:4])
+	}
+
+	// Extract which network this invoice is meant for.
+	var net *chaincfg.Params
+	switch hrp[2:4] {
```

Comment:
Hmm, it definitely will. BIP-173 only specifies "bc" and "tb", but it indeed looks like litecoin is going to use "ltc" as prefix. This switch should probably be changed in order to support more coins than bitcoin. Currently it is not that easy, since there is no mapping from `prefix->chaincfg.Params` available.

On way could be to do something like
```go
if strings.HasPrefix(hrp[2:], chaincfg.MainNetParams.Bech32HRPSegwit) {
	net = &chaincfg.MainNetParams
} else if strings.HasPrefix(hrp[2:], chaincfg.TestNet3Params.Bech32HRPSegwit) {
	net = &chaincfg.TestNet3Params
} else if strings.HasPrefix(hrp[2:], chaincfg.SimNetParams.Bech32HRPSegwit) {
	net = &chaincfg.SimNetParams
} else {
	return nil, fmt.Errorf("unknown network")
}
```
What do you think, or other suggestions?

Reply:
Ended up doing this for now, lemme know if better idea.